### PR TITLE
[dynamo] add SymNode bitwise and/or

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138777

Fixes [T203472723](https://www.internalfb.com/intern/tasks/?t=203472723)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ezyang @SherlockNoMad @EikanWang @wenzhe-nrv @penguinwu @bobrenjc93 @voznesenskym @Guobing-Chen @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/utils/_sympy/value_ranges.py [hunk 1]
```diff
@@ -499,6 +499,32 @@ def or_(a, b):
...
+        if upper == 0:
+            upper = 0
+        elif upper > 0 and upper != int_oo:
+            # If both upper bounds are positive, then the largest
+            # possible value is 01...1, so we need to find
+            # next largest power of 2 (exclusive), minus 1
+            upper = 2 ** (math.floor(math.log2(upper)) + 1) - 1
+        elif upper < 0:
+            upper = -1
+        return ValueRanges(min(a.lower, b.lower), upper)
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:526]: imma trust the tests on these, they're pretty spooky haha

File: torch/fx/experimental/sym_node.py [hunk 2]
```diff
@@ -1565,9 +1589,12 @@ def round_magic_impl(self, ndigits=None):
 
         setattr(user_type, f"__{method}__", round_magic_impl)
     else:
-        setattr(user_type, f"__{method}__", binary_magic_impl)
+        method_name = method
+        if method in bitwise_ops:
+            method_name = method[len("bitwise_") :]
```
Comment (ezyang) [inline torch/fx/experimental/sym_node.py:1594]: annoying lol

**Comment by ezyang** [comment]
This seems fine but it looks like you have to fix some tests, and I would appreciate more thorough testing on the value ranges logic

**Comment by williamwen42** [comment]
Added more tests, support some bitwise ops in `torch.fx.experimental.validator.z3op`.

File: torch/fx/experimental/validator.py [hunk 3]
```diff
@@ -277,12 +277,37 @@ def wrapper(*args):
 
             return wrapper
 
+        # Like `lift`, but for bitwise ops. We need to convert to/from BitVec
+        # in order to use z3 bitwise ops. We assume that integers are 64 bit.
```
Comment (williamwen42) [inline torch/fx/experimental/validator.py:281]: Safe to assume?

Comment (ezyang) [reply torch/fx/experimental/validator.py:281]: We don't currently make this assumption I think, but it might also be harmless here. cc @ysiraichi

Comment (ysiraichi) [reply torch/fx/experimental/validator.py:281]: Probably safe to assume. But, not sure this will play well with the other operations. Could you add a test where there is a, for example, bitwise operation followed by, say, `mod`?

File: test/test_fx_experimental.py [hunk 4]
```diff
@@ -563,8 +566,18 @@ def test_conv_bn_fusion_mixed_dtype(self):
...
+                    16,
+                    kernel_size=3,
+                    stride=1,
+                    padding=1,
+                    bias=False,
+                    dtype=torch.bfloat16,
+                )
+                self.bn = torch.nn.BatchNorm2d(
+                    16, eps=0.001, momentum=0.1, affine=True, track_running_stats=True
+                )
```
Comment (ezyang) [inline test/test_fx_experimental.py:580]: bah why you reformat unrelated code

**Review by ezyang — CHANGES_REQUESTED** [review]
stack the reformat in a stack diff before this one, don't do that to your reviewers

File: test/test_fx_experimental.py (hunk 4 above)
Comment (williamwen42) [reply test/test_fx_experimental.py:580]: Hmm not sure what happened here - I think some linter was automatically applied.

Comment (ezyang) [reply test/test_fx_experimental.py:580]: I have found the usual explanation  is you pyfmt version doesn't match the repo. Lintrunner init will fix it

**Comment by williamwen42** [comment]
@ysiraichi any ideas why the inductor benchmark is getting a z3 type error? I can't make a minimal repro for this and it seems to require aot_eager backend.

**Comment by ysiraichi** [comment]
Tried looking into it, but not really sure what's happening there. It's odd that it's declaring as `(declare-fun int2bv (Bool) (_ BitVec 64))` with `Bool` instead of `Int`.

Another thing: after introducing BV with Int_Reals, maybe `QF_NRA` logic won't be enough for finding a solution. We would need to change into something that supports BV.

**Comment by ezyang** [comment]
TBH, I'm not that interested in trying to get the BV theory to work, I'm OK with less good TV for bitwise and/or

File: torch/fx/experimental/validator.py [hunk 5]
```diff
@@ -277,12 +277,42 @@ def wrapper(*args):
...
+            def wrapper(*args):
+                # Lifts the arguments into a list of Z3 inhabitants.
+                if len(args) == 1 and isinstance(args[0], (list, tuple)):
+                    wrapped_args = (tuple(wrap(a) for a in args[0]),)
+                else:
+                    wrapped_args = tuple(wrap(a) for a in args)
+                # Run the function on the Z3 expressions.
+                return z3.BV2Int(func(*wrapped_args))
+
+            return wrapper
```
Comment (ysiraichi) [inline torch/fx/experimental/validator.py:306]: What do you think about reusing the code in the `lift` function? Both `wrap` and `wrapper` look pretty similar to me. Then, you could just wrap them with a bitwise wrapper.

File: torch/fx/experimental/validator.py [hunk 6]
```diff
@@ -510,9 +540,11 @@ def to_z3_boolean_expr(self, e: sympy.Basic) -> z3.BoolRef:
             ), f"expected boolean expression. Got: {z3expr}"
             return z3expr
 
-        def add_source_expr(self, e: z3.BoolRef) -> None:
+        def add_source_expr(self, e: Union[z3.BoolRef, z3.ArithRef]) -> None:
```
Comment (ysiraichi) [inline torch/fx/experimental/validator.py:543]: Why is this needed?

File: torch/fx/experimental/validator.py [hunk 7]
```diff
@@ -277,12 +277,42 @@ def wrapper(*args):
...
+
+            return wrapper
+
         ops = _Z3Ops(validator)
         replacement_map = {
             # Operator module.
             operator.not_: lift(z3.Not),
-            operator.and_: lift(z3.And),
-            operator.or_: lift(z3.Or),
+            operator.and_: bitwise_lift(operator.and_),
```
Comment (ysiraichi) [inline torch/fx/experimental/validator.py:312]: Maybe for operators `and_` and `or_` we could check if the arguments are bool, in order to make it easier for Z3 to solve, instead of doing logical operations in BV.

File: torch/fx/experimental/validator.py (hunk 5 above)
Comment (williamwen42) [reply torch/fx/experimental/validator.py:306]: Yep, the 2 functions became more similar after some changes I made.

File: torch/fx/experimental/validator.py (hunk 6 above)
Comment (williamwen42) [reply torch/fx/experimental/validator.py:543]: Boolean and integer bitwise ops go through the same implementation, which now usually returns an integer (except for that weird case with the failing benchmark that I'm still aiming to repro). But this part of the code expects a boolean expression, so I changed this function to accept integer/arith expression, with intent to convert into a boolean expression.

File: torch/fx/experimental/validator.py (hunk 7 above)
Comment (williamwen42) [reply torch/fx/experimental/validator.py:312]: I can try having a separate case when both arguments are bool.

Comment (ezyang) [reply torch/fx/experimental/validator.py:312]: Yeah, I am worried that the change here is going to cause problems for regular bool conjunction/disjunction. Ideally, we are able to distinguish if something is a logical and / logical or versus bitwise and / bitwise or. If we've been abusing `operator.and_` and `operator.or_` for both cases, we may need to add a dedicated logical and/or that is its dedicated operator.

This is especially pernicious because bool is a subtype of int and they are NOT compatible.

File: torch/utils/_sympy/value_ranges.py [hunk 8]
```diff
@@ -500,6 +500,53 @@ def or_(a, b):
...
+    @staticmethod
+    def _bool_to_int(x):
+        if x.is_singleton():
+            return ValueRanges.wrap(sympy.Integer(1 if x.lower else 0))
+        else:
+            return ValueRanges(sympy.Integer(0), sympy.Integer(1))
+
+    @staticmethod
+    def _bit_length(n):
+        return len(bin(n).lstrip("-0b"))
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:512]: lol wtf is this. Do better

File: torch/utils/_sympy/value_ranges.py [hunk 9]
```diff
@@ -500,6 +500,53 @@ def or_(a, b):
...
+
+    @classmethod
+    def bitwise_and(cls, a, b):
+        a, b = ValueRanges.wrap(a), ValueRanges.wrap(b)
+        if a.is_bool:
+            a = cls._bool_to_int(a)
+        if b.is_bool:
+            b = cls._bool_to_int(b)
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:520]: This is not right. The problem is that bitwise operations on booleans are defined to return boolean:

```
>>> True | False
True
```

so you are obligated to return a boolean ValueRanges when both arguments are boolean. We are also a little loosey-goosey about this, but my general preference is to insert explicit coercions when they would have happened, e.g., `2 | True` should become `2 | bool_to_int(True)`

File: torch/utils/_sympy/value_ranges.py (hunk 8 above)
Comment (williamwen42) [reply torch/utils/_sympy/value_ranges.py:512]: I took this from https://docs.python.org/3/library/stdtypes.html#int.bit_length because sympy Integer doesn't implement bit_length. Otherwise, I need to write a loop, or use log2.

**Comment by williamwen42** [comment]
Changed all sites with an explicit bool -> int conversion to use bool op if both arguments are bool.

**Comment by ysiraichi** [comment]
Thank you for addressing my reviews. There's still one thing I would like to see in this PR: a validator test that shows we are able to find a model that satisfies our validator problem, i.e. yields `sat` when calling `Solver.check()`.

File: torch/utils/_sympy/value_ranges.py (hunk 8 above)
Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:512]: What's the input type here? If it's an int you can just use `bit_length` method on it?

Comment (williamwen42) [reply torch/utils/_sympy/value_ranges.py:512]: It could be sympy Integer, which doesn't have bit_length.

Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:512]: If it is guaranteed to be sympy.Integer you can convert it to int

File: torch/utils/_sympy/value_ranges.py [hunk 10]
```diff
@@ -500,6 +500,57 @@ def or_(a, b):
...
+        if upper == 0:
+            upper = 0
+        elif upper > 0 and upper != int_oo:
+            # If both upper bounds are positive, then the largest
+            # possible value is 01...1, so we need to find
+            # next largest power of 2 (exclusive), minus 1
+            upper = (1 << cls._bit_length(upper)) - 1
+        elif upper < 0:
+            upper = -1
+        return ValueRanges(min(a.lower, b.lower), upper)
```
Comment (ezyang) [inline torch/utils/_sympy/value_ranges.py:552]: This bound does not seem tight. Counterexample: Suppose we have [0b101, inf] and [0b010, inf]. Then the minimum per analysis here is 0b010.  But it is impossible to attain this minimum: 0b101 | 0b010 = 0b111, and any larger value for 0b101 or 0b010 would cause the X in 0bX00 (or a higher bit) to be set.

Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:552]: I have a broader question, which is do we *really* need value range analysis here? Like can we not just say that bitwise operations preserve sign and leave it at that?

Comment (williamwen42) [reply torch/utils/_sympy/value_ranges.py:552]: We could certainly just say that bitwise ops preserve sign - I don't completely understand the purpose of ValueRanges, so the current implementation is a best effort bound that is not guaranteed to be tight.

Comment (ezyang) [reply torch/utils/_sympy/value_ranges.py:548]: ValueRanges is used to make determinations if we statically know things to true. It is designed to be a simple but powerful reasoning mechanism. For most size variables we know their range is `[2, inf]`. So we can easily conclude that e.g., `2 * s0 != 0`.

**Review by ezyang — APPROVED** [review]
whatever, if it's buggy, we'll figure it out lol

**Comment by ezyang** [comment]
So I believe the fix is you need to add entries to SYMPY_INTERP in symbolic_shapes.py but given that this is something that we've run into several times it would also be nice to have a test that this dict has enough entries so when new ops are added we don't forget too.

**Comment by ezyang** [comment]
> I'm also looking at https://github.com/pytorch/pytorch/pull/137980 and saw that you didn't need to add OpaqueUnaryFn_log2 to SYMPY_INTERP

This probably also is a bug

**Comment by williamwen42** [comment]
Added a test that evaluates bitwise and/or in a guard - will followup with "it would also be nice to have a test that this dict has enough entries so when new ops are added we don't forget too" and fixing log2 in a separate PR.

File: torch/fx/experimental/symbolic_shapes.py [hunk 11]
```diff
@@ -2025,6 +2025,8 @@ def cast_symbool_to_symint_guardless(
     "OpaqueUnaryFn_tanh": math.tanh,
     "OpaqueUnaryFn_atan": math.atan,
     "OpaqueUnaryFn_sqrt": math.sqrt,
+    "BitwiseFn_bitwise_and": operator.and_,
+    "BitwiseFn_bitwise_or": operator.or_,
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2029]: endorsed
